### PR TITLE
amber: Improved package.py and added version 18

### PR DESCRIPTION
**Pull request by hseara**
- Added amber 18 with ambertools 19
- Added all available patches
- Added +update variant to use the self update
- Added +openmp variant to get openmp optomizations
- Added +x11 variant when possible
- Splitted amber 16 and 18 dependencies
- We now detect the copiler type and compile accordingly
- Added cray variant which is a bit special (untested)
- Improved detection of possible cuda versions
- All compilation optimizations +mpi +openmp +cuda are compatible
- Updated to use setup_build_environment(), setup_run_environment()

### CODE REVIEW

**Comment by adamjstewart** [comment]
> Updated to use setup_build_environment(), setup_run_environment()

I recently merged a PR that does this for every package in Spack, so you'll need to rebase/resolve merge conflicts.

**Comment by hseara** [comment]
Please do not merge yet. I am still adding some missing dependencies.

**Comment by adamjstewart** [comment]
The `dealii +tbb` stuff was already merged in a separate PR, so you might need to rebase again.

**Comment by hseara** [comment]
I am done from my side. Unless there is more job you need from my side, feel free to merge.

File: var/spack/repos/builtin/packages/amber/package.py [hunk 1]
```diff
@@ -17,63 +18,177 @@ class Amber(Package, CudaPackage):
...
+            default=True)
+    variant('openmp', description='Use OpenMP pragmas to parallelize',
+            default=False)
+    variant('python', description='Build python applications',
+            default=False)
+    variant('x11', description='Build programs that require X11',
+            default=False)
+    variant('update', description='Update the sources prior compilation',
+            default=False)
+    variant('cray', description='Compiling in a cray server',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/amber/package.py]: A better way is to check `spec.satisfies('platform=cray')`, then you don't need an extra variant.

File: var/spack/repos/builtin/packages/amber/package.py [hunk 2]
```diff
@@ -17,63 +18,177 @@ class Amber(Package, CudaPackage):
...
-    depends_on('cuda@7.5.18', when='+cuda')
-
-    depends_on('netcdf-fortran')
-    depends_on('python+tkinter@2.7:2.8', type=('build', 'run'))
-    depends_on('py-numpy', type=('build', 'run'))
-    depends_on('py-scipy', type=('build', 'run'))
-    depends_on('py-matplotlib@:2.9', type=('build', 'run'))
     depends_on('zlib')
+    depends_on('flex')
+    depends_on('bison')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/amber/package.py]: Flex and bison are usually `type='build'` dependencies only

File: var/spack/repos/builtin/packages/amber/package.py [hunk 3]
```diff
@@ -17,63 +18,177 @@ class Amber(Package, CudaPackage):
...
-        sp_dir = join_path(self.prefix, 'python2.7/site-packages')
-
-        env.set('AMBERHOME', self.prefix)
-        env.prepend_path('PYTHONPATH', sp_dir)
+    # Cuda dependencies
+    depends_on('cuda@:10.1.243', when='@18:+cuda')
+    depends_on('cuda@7.5.18', when='@:16+cuda')
+
+    # conflicts
+    conflicts('+x11', when='+cray',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/amber/package.py]:
```suggestion
    conflicts('+x11', when='platform=cray,
```

File: var/spack/repos/builtin/packages/amber/package.py [hunk 4]
```diff
@@ -17,63 +18,177 @@ class Amber(Package, CudaPackage):
...
-
+        # The resource command does not allow us to expand the package in the
+        # root stage folder as required, as it already contains files. Here we
+        # install AmberTools where it should be, which results in 3 copies of
+        # the  ambertools (~9 GB). This has to be improved in the future.
+        install_tree('ambertools_tmpdir', '.')
+        shutil.rmtree(join_path(self.stage.source_path, 'ambertools_tmpdir'))
+
+        # Select compiler style
+        if self.spec.satisfies('+cray'):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/amber/package.py]:
```suggestion
        if self.spec.satisfies('%cce'):
```

File: var/spack/repos/builtin/packages/amber/package.py [hunk 5]
```diff
@@ -17,63 +18,177 @@ class Amber(Package, CudaPackage):
...
+            make('install')
 
         # just install everything that was built
         install_tree('.', prefix)
+
+    def setup_run_environment(self, env):
+        env.set('AMBER_PREFIX', self.prefix)
+        env.set('AMBERHOME', self.prefix)
+        env.prepend_path('PATH', self.prefix.bin)
+        env.prepend_path('LD_LIBRARY_PATH', self.prefix.lib)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/amber/package.py]: These two aren't needed, they are automatically added to the module file.

**Comment by hseara** [comment]
I am not sure how dealii ended up in my pull request. It was certainly not my intent. Anyway, I fixed some flake8 problems with dealii, that were making my checks to fail. Please let me know if the package is not ready for merging.

File: var/spack/repos/builtin/packages/amber/package.py (hunk 5 above)
Comment (hseara) [reply var/spack/repos/builtin/packages/amber/package.py]: All suggestions accepeted and commited.

File: var/spack/repos/builtin/packages/amber/package.py (hunk 4 above)
Comment (hseara) [reply var/spack/repos/builtin/packages/amber/package.py]: Done

File: var/spack/repos/builtin/packages/amber/package.py [hunk 6]
```diff
@@ -17,63 +18,173 @@ class Amber(Package, CudaPackage):
...
+    for ver, num, checksum in patches:
+        patch_url_str = 'https://ambermd.org/bugfixes/{0}.0/update.{1}'
+        patch(patch_url_str.format(ver, num),
+              sha256=checksum, level=0, when='@{0}'.format(ver))
+
+    variant('mpi', description='Build MPI executables',
+            default=True)
+    variant('openmp', description='Use OpenMP pragmas to parallelize',
+            default=False)
+    variant('python', description='Build python applications',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/amber/package.py]: This variant doesn’t seem to be used right now?

Comment (hseara) [reply var/spack/repos/builtin/packages/amber/package.py]: I have removed it for now. This variant will be used in the future, but currently, I have problems making the +python compilation to work. Therefore, I left it out from this initial commit.

**Pull request merged**